src/client/websocket/WebSocketManager.js

```diff
@@ -37,52 +77,148 @@ class WebSocketManager extends EventEmitter {
+      const shard = new WebSocketShard(this, item);
+      this.shards[item] = shard;
+      shard.once(Events.READY, () => {
+        this.spawning = false;
+        this.client.setTimeout(() => this.spawn(), 5000);
+      });
+    } else if (item instanceof WebSocketShard) {
+      item.reconnect();
     }
-    return this.connection.destroy();
   }
 
   /**
-   * Send a packet on the available WebSocket.
-   * @param {Object} packet Packet to send
+   * Creates a connection to a gateway.
+   * @param {string} [gateway] The gateway to connect to
    * @returns {void}
    */
-  send(packet) {
-    if (!this.connection) {
-      this.debug('No connection to websocket');
-      return;
+  connect(gateway = this.gateway) {
+    this.gateway = gateway;
+
+    if (typeof this.client.options.shards === 'number') {
+      this.debug('Spawning 1 shard');
+      this.spawn(this.client.options.shards);
+    } else if (Array.isArray(this.client.options.shards)) {
+      this.debug(`Spawning ${this.client.options.shards.length} shards`);
+      for (let i = 0; i < this.client.options.shards.length; i++) {
+        this.spawn(this.client.options.shards[i]);
+      }
+    } else {
+      this.debug(`Spawning ${this.client.options.shardCount} shards`);
+      for (let i = 0; i < this.client.options.shardCount; i++) {
+        this.spawn(i);
+      }
     }
-    this.connection.send(packet);
   }
 
   /**
-   * Connects the client to a gateway.
-   * @param {string} gateway The gateway to connect to
+   * Processes a packet and queues it if this WebSocketManager is not ready.
+   * @param {Object} packet The packet to be handled
+   * @param {WebSocketShard} shard The shard that will handle this packet
    * @returns {boolean}
    */
-  connect(gateway) {
-    if (!this.connection) {
-      this.connection = new WebSocketConnection(this, gateway);
-      return true;
-    }
-    switch (this.connection.status) {
-      case Status.IDLE:
-      case Status.DISCONNECTED:
-        this.connection.connect(gateway, 5500);
-        return true;
-      default:
-        this.debug(`Couldn't connect to ${gateway} as the websocket is at state ${this.connection.status}`);
+  handlePacket(packet, shard) {
+    if (packet && this.status !== Status.READY) {
+      if (BeforeReadyWhitelist.indexOf(packet.t) === -1) {
+        this.packetQueue.push({ packet, shardID: shard.id });
         return false;
+      }
+    }
+
+    if (this.packetQueue.length) {
+      const item = this.packetQueue.shift();
+      this.client.setImmediate(() => {
+        this.handlePacket(item.packet, this.shards[item.shardID]);
+      });
+    }
+
+    if (packet && PacketHandlers[packet.t]) {
+      PacketHandlers[packet.t](this.client, packet, shard);
+    }
+
+    return false;
+  }
+
+  /**
+   * Checks whether the client is ready to be marked as ready.
+   * @returns {void}
+   */
+  checkReady() {
+    if (!(this.shards.filter(s => !!s).length === this.client.options.shardCount) ||
```

Comment:
Shouldn't that be `Boolean(s)` and is the conversion to an explicit boolean even necessary?